`publishApk` fails on clean builds if Gradle configuration cache is enabled

> [!NOTE]
> Claude Opus 4.7 made the original writeup. I have rewritten this report for brevity and clarity.

### Describe the bug

Executing `./gradlew clean publishPlayReleaseApk` with `org.gradle.configuration-cache=true` fails with `Cannot query the value of this provider because it has no value available.`

<details>

<summary>Build output/stack trace</summary>

```
> Task :AnkiDroid:publishPlayReleaseApk FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':AnkiDroid:publishPlayReleaseApk'.
> Cannot query the value of this provider because it has no value available.

Caused by: org.gradle.api.internal.provider.MissingValueException: Cannot query the value of this provider because it has no value available.
    at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateOwnPresentValue(AbstractMinimalProvider.java:82)
    at org.gradle.api.internal.provider.AbstractMinimalProvider.get(AbstractMinimalProvider.java:100)
    at org.gradle.api.internal.provider.ProviderResolutionStrategy$2.resolve(ProviderResolutionStrategy.java:32)
    at org.gradle.api.internal.file.collections.ProviderBackedFileCollection.visitChildren(ProviderBackedFileCollection.java:63)
    ...
    at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection.calculateFinalizedValue(DefaultConfigurableFileCollection.java:423)
    at org.gradle.api.internal.tasks.PropertyFileCollection.visitChildren(PropertyFileCollection.java:48)
    at org.gradle.internal.fingerprint.impl.DefaultFileCollectionSnapshotter.snapshot(DefaultFileCollectionSnapshotter.java:48)
    at org.gradle.internal.execution.impl.DefaultInputFingerprinter$InputCollectingVisitor.visitInputFileProperty(DefaultInputFingerprinter.java:151)
    at org.gradle.internal.execution.steps.SkipEmptyMutableWorkStep$2.visitInputFileProperty(SkipEmptyMutableWorkStep.java:190)
```

</details>

In CI: https://github.com/ankidroid/Anki-Android/actions/runs/24752127464/job/72417332528

### How To Reproduce

<details>

<summary>Create file: ~/src/AnkiDroid-GCP-Publish-Credentials.json</summary>

```json
{
  "type": "service_account",
  "project_id": "ankidroid-config-cache-diagnosis",
  "private_key_id": "0000000000000000000000000000000000000000",
  "private_key": "-----BEGIN PRIVATE KEY-----\nDUMMY-KEY-FOR-LOCAL-CONFIG-CACHE-DIAGNOSIS-ONLY-NOT-A-REAL-CREDENTIAL\n-----END PRIVATE KEY-----\n",
  "client_email": "dummy@ankidroid-config-cache-diagnosis.iam.gserviceaccount.com",
  "client_id": "000000000000000000000",
  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
  "token_uri": "https://oauth2.googleapis.com/token",
  "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
  "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/dummy"
}
```

</details>

```bash
mkdir 20850 && cd 20850
git clone https://github.com/ankidroid/Anki-Android.git && cd Anki-Android
git checkout 22fe96c79b80c5db1339fc2104f2048f65ec0118
# MANUAL STEP: setup `AnkiDroid-GCP-Publish-Credentials.json` (above) 
./gradlew clean :AnkiDroid:publishPlayReleaseApk -PbuildTime=$(date +%s000) --stacktrace
# `MissingValueException` at `SkipEmptyMutableWorkStep$2.visitInputFileProperty`
./gradlew :AnkiDroid:publishPlayReleaseApk -PbuildTime=$(date +%s000) --stacktrace
# Expected failure: Credential parsing may have failed.
```

**EDIT:** After reproduction issues, I could reproduce by deleting `.gradle`, `build` and `AnkiDroid/build` and `/tmp/fresh-gradle-**` folders and and running: 

```bash
GRADLE_USER_HOME=/tmp/fresh-gradle-$$ ./gradlew --stop
GRADLE_USER_HOME=/tmp/fresh-gradle-$$ ./gradlew clean :AnkiDroid:publishPlayReleaseApk -PbuildTime=$(date +%s000) --stacktrace --no-build-cache --no-parallel -Dorg.gradle.configuration-cache.parallel=false -Dorg.gradle.workers.max=1
```

#### Versions

- Gradle Play Publisher: 4.0.0
- Gradle Wrapper: 9.4.1
- Android Gradle Plugin: 9.0.1

#### Tasks executed

`./gradlew :AnkiDroid:publishPlayReleaseApk -PbuildTime=$(date +%s000) --stacktrace`

### Expected behavior

No error in production. In the test script:

> Credential parsing may have failed. Ensure credential files supplied in the DSL contain valid JSON and/or the ANDROID_PUBLISHER_CREDENTIALS envvar contains valid JSON (not a file path).

### Additional context

https://github.com/Triple-T/gradle-play-publisher/blob/2684cef311c1bbb4b78fc81dba4c29e8b358b393/play/plugin/src/main/kotlin/com/github/triplet/gradle/play/PlayPublisherPlugin.kt#L215-L227

Due to the configuration cache, the following code is run in the store phase, before `output-metadata.json` is produced by `packagePlayRelease`. `null` is produced, but deferred by `sneakyNull()`.

https://github.com/Triple-T/gradle-play-publisher/blob/2684cef311c1bbb4b78fc81dba4c29e8b358b393/play/plugin/src/main/kotlin/com/github/triplet/gradle/play/PlayPublisherPlugin.kt#L225-L226

Accessing `apks` then fails

https://github.com/Triple-T/gradle-play-publisher/blob/2684cef311c1bbb4b78fc81dba4c29e8b358b393/play/plugin/src/main/kotlin/com/github/triplet/gradle/play/PlayPublisherPlugin.kt#L486

https://github.com/Triple-T/gradle-play-publisher/blob/2684cef311c1bbb4b78fc81dba4c29e8b358b393/play/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/PublishApk.kt#L39-L42


<details>

<summary>Explanation - Claude Opus 4.7</summary>

This was directly verified by running `apkProvider.map { dir -> loader.load(dir)?.elements?.collect { it.outputFile } }.get()` from an init script that depends on `packagePlayRelease`. The lambda still fires twice during cache store with `loaded=NULL`, then `findApkFiles.get()` throws `MissingValueException` at task action time — *while a separate, freshly-built call to `loader.load(dir)` against the same directory at the same point in time succeeds and returns 4 elements*. The loader works; the memoized Provider does not.

</details>

--- a/play/plugin/src/main/kotlin/com/github/triplet/gradle/play/PlayPublisherPlugin.kt
+++ b/play/plugin/src/main/kotlin/com/github/triplet/gradle/play/PlayPublisherPlugin.kt
@@ -51,6 +51,7 @@ import com.github.triplet.gradle.play.tasks.internal.WriteTrackLifecycleTask
 import org.gradle.api.Plugin
 import org.gradle.api.Project
 import org.gradle.api.Task
+import org.gradle.api.file.FileCollection
 import org.gradle.api.plugins.ExtensionAware
 import org.gradle.api.provider.Provider
 import org.gradle.api.services.BuildServiceRegistration
@@ -212,19 +213,15 @@ internal abstract class PlayPublisherPlugin @Inject constructor(
                 return@v
             }
 
-            fun findApkFiles(): Provider<List<String>> = extension.artifactDir.map {
-                val customDir = it.asFile
-                if (customDir.isFile && customDir.extension == "apk") {
-                    listOf(it.asFile.absolutePath)
+            // #1187: Only return the sources. APKs must be resolved by ApkSpec.resolve()
+            fun findApkFileSources(): Provider<FileCollection> = extension.artifactDir.map { dir ->
+                val file = dir.asFile
+                if (file.isFile && file.extension == "apk") {
+                    project.objects.fileCollection().from(file)
                 } else {
-                    it.asFileTree.matching {
-                        include("*.apk")
-                    }.map { it.absolutePath }
+                    dir.asFileTree.matching { include("*.apk") }
                 }
-            }.orElse(variant.artifacts.get(SingleArtifact.APK).map {
-                variant.artifacts.getBuiltArtifactsLoader().load(it)
-                        ?.elements?.map { it.outputFile }.sneakyNull()
-            })
+            }.orElse(variant.artifacts.get(SingleArtifact.APK).map { it.asFileTree })
 
             fun findBundleFiles(): Provider<List<String>> = extension.artifactDir.map {
                 val customDir = it.asFile
@@ -284,7 +281,9 @@ internal abstract class PlayPublisherPlugin @Inject constructor(
                     arrayOf(extension, executionDir),
             ) {
                 bindApi(api)
-                apks.from(findApkFiles())
+                apkSpec.sources.from(findApkFileSources())
+                apkSpec.artifactDir.set(extension.artifactDir)
+                apkSpec.builtArtifactsLoader.set(variant.artifacts.getBuiltArtifactsLoader())
                 outputDirectory.set(project.layout.buildDirectory.dir(
                         "outputs/internal-sharing/apk/${variant.name}"))
 
@@ -483,7 +482,9 @@ internal abstract class PlayPublisherPlugin @Inject constructor(
                     arrayOf(extension, executionDir),
             ) {
                 configureInputs()
-                apks.from(findApkFiles())
+                apkSpec.sources.from(findApkFileSources())
+                apkSpec.artifactDir.set(extension.artifactDir)
+                apkSpec.builtArtifactsLoader.set(variant.artifacts.getBuiltArtifactsLoader())
                 mappingFiles.from(extension.artifactDir.map { customDir ->
                     project.objects.fileCollection().from(customDir.asFileTree.matching {
                         include("mapping.txt", "*.mapping.txt")

--- a/play/plugin/src/main/kotlin/com/github/triplet/gradle/play/internal/ApkSpec.kt
+++ b/play/plugin/src/main/kotlin/com/github/triplet/gradle/play/internal/ApkSpec.kt
@@ -1,0 +1,34 @@
+package com.github.triplet.gradle.play.internal
+
+import com.android.build.api.variant.BuiltArtifactsLoader
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.InputFiles
+import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.PathSensitive
+import org.gradle.api.tasks.PathSensitivity
+import org.gradle.api.tasks.SkipWhenEmpty
+import java.io.File
+
+internal abstract class ApkSpec {
+    @get:PathSensitive(PathSensitivity.RELATIVE)
+    @get:SkipWhenEmpty
+    @get:InputFiles
+    abstract val sources: ConfigurableFileCollection
+
+    @get:Internal
+    abstract val artifactDir: DirectoryProperty
+
+    @get:Internal
+    abstract val builtArtifactsLoader: Property<BuiltArtifactsLoader>
+
+    // loader.load() returns null during configuration-cache store
+    // So resolve APKs at task-action time (#1187)
+    fun resolve(): List<File> = if (artifactDir.isPresent) {
+        sources.files.toList()
+    } else {
+        builtArtifactsLoader.get().load(sources)?.elements?.map { File(it.outputFile) }
+                ?: error("output-metadata.json not found in ${sources.asPath}")
+    }
+}

--- a/play/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/PublishApk.kt
+++ b/play/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/PublishApk.kt
@@ -2,6 +2,7 @@ package com.github.triplet.gradle.play.tasks
 
 import com.github.triplet.gradle.common.utils.safeCreateNewFile
 import com.github.triplet.gradle.play.PlayPublisherExtension
+import com.github.triplet.gradle.play.internal.ApkSpec
 import com.github.triplet.gradle.play.tasks.internal.CliOptionsImpl
 import com.github.triplet.gradle.play.tasks.internal.PublishArtifactTaskBase
 import com.github.triplet.gradle.play.tasks.internal.PublishableTrackExtensionOptions
@@ -16,6 +17,7 @@ import org.gradle.api.file.RegularFileProperty
 import org.gradle.api.provider.ListProperty
 import org.gradle.api.tasks.InputFile
 import org.gradle.api.tasks.InputFiles
+import org.gradle.api.tasks.Nested
 import org.gradle.api.tasks.Optional
 import org.gradle.api.tasks.OutputDirectory
 import org.gradle.api.tasks.PathSensitive
@@ -36,10 +38,8 @@ internal abstract class PublishApk @Inject constructor(
         private val executor: WorkerExecutor,
 ) : PublishArtifactTaskBase(extension),
         PublishableTrackExtensionOptions by CliOptionsImpl(extension, executionDir) {
-    @get:PathSensitive(PathSensitivity.RELATIVE)
-    @get:SkipWhenEmpty
-    @get:InputFiles
-    internal abstract val apks: ConfigurableFileCollection
+    @get:Nested
+    internal abstract val apkSpec: ApkSpec
 
     @get:PathSensitive(PathSensitivity.RELATIVE)
     @get:SkipWhenEmpty
@@ -63,7 +63,7 @@ internal abstract class PublishApk @Inject constructor(
         executor.noIsolation().submit(Processor::class) {
             paramsForBase(this)
 
-            apkFiles.set(apks)
+            apkFiles.set(apkSpec.resolve())
             deobfuscationFiles.from(mappingFiles)
             debugSymbolsFile.set(nativeDebugSymbols)
             uploadResults.set(temporaryDir)

--- a/play/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/PublishInternalSharingApk.kt
+++ b/play/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/PublishInternalSharingApk.kt
@@ -1,22 +1,19 @@
 package com.github.triplet.gradle.play.tasks
 
 import com.github.triplet.gradle.play.PlayPublisherExtension
+import com.github.triplet.gradle.play.internal.ApkSpec
 import com.github.triplet.gradle.play.tasks.internal.ArtifactExtensionOptions
 import com.github.triplet.gradle.play.tasks.internal.CliOptionsImpl
 import com.github.triplet.gradle.play.tasks.internal.PublishTaskBase
 import com.github.triplet.gradle.play.tasks.internal.workers.PlayWorkerBase
 import com.github.triplet.gradle.play.tasks.internal.workers.copy
 import com.github.triplet.gradle.play.tasks.internal.workers.paramsForBase
-import org.gradle.api.file.ConfigurableFileCollection
 import org.gradle.api.file.Directory
 import org.gradle.api.file.DirectoryProperty
 import org.gradle.api.file.RegularFileProperty
 import org.gradle.api.provider.ListProperty
-import org.gradle.api.tasks.InputFiles
+import org.gradle.api.tasks.Nested
 import org.gradle.api.tasks.OutputDirectory
-import org.gradle.api.tasks.PathSensitive
-import org.gradle.api.tasks.PathSensitivity
-import org.gradle.api.tasks.SkipWhenEmpty
 import org.gradle.api.tasks.TaskAction
 import org.gradle.kotlin.dsl.submit
 import org.gradle.work.DisableCachingByDefault
@@ -31,10 +28,8 @@ internal abstract class PublishInternalSharingApk @Inject constructor(
         private val executor: WorkerExecutor,
 ) : PublishTaskBase(extension),
         ArtifactExtensionOptions by CliOptionsImpl(extension, executionDir) {
-    @get:PathSensitive(PathSensitivity.RELATIVE)
-    @get:SkipWhenEmpty
-    @get:InputFiles
-    internal abstract val apks: ConfigurableFileCollection
+    @get:Nested
+    internal abstract val apkSpec: ApkSpec
 
     @get:OutputDirectory
     abstract val outputDirectory: DirectoryProperty
@@ -44,7 +39,7 @@ internal abstract class PublishInternalSharingApk @Inject constructor(
         executor.noIsolation().submit(Processor::class) {
             paramsForBase(this)
 
-            apkFiles.set(apks)
+            apkFiles.set(apkSpec.resolve())
             outputDir.set(outputDirectory)
         }
     }

--- a/play/plugin/src/test/kotlin/com/github/triplet/gradle/play/helpers/IntegrationTestBase.kt
+++ b/play/plugin/src/test/kotlin/com/github/triplet/gradle/play/helpers/IntegrationTestBase.kt
@@ -56,6 +56,9 @@ abstract class IntegrationTestBase(
     override fun executeExpectingFailure(config: String, vararg tasks: String) =
             execute(config, true, *tasks)
 
+    fun execute(config: String, vararg tasks: String, extraArgs: List<String>) =
+            execute(config, false, *tasks, extraArgs = extraArgs)
+
     override fun executeGradle(
             expectFailure: Boolean,
             block: GradleRunner.() -> Unit,
@@ -100,6 +103,7 @@ abstract class IntegrationTestBase(
             config: String,
             expectFailure: Boolean,
             vararg tasks: String,
+            extraArgs: List<String> = emptyList(),
     ): BuildResult {
         val buildCacheDir = File(tempDir, "gradle").escaped()
 
@@ -153,7 +157,7 @@ abstract class IntegrationTestBase(
         """)
 
         return executeGradle(expectFailure) {
-            withArguments("--build-cache", *tasks)
+            withArguments("--build-cache", *extraArgs.toTypedArray(), *tasks)
         }
     }
 

--- a/play/plugin/src/test/kotlin/com/github/triplet/gradle/play/tasks/PublishApkIntegrationTest.kt
+++ b/play/plugin/src/test/kotlin/com/github/triplet/gradle/play/tasks/PublishApkIntegrationTest.kt
@@ -51,6 +51,17 @@ class PublishApkIntegrationTest : IntegrationTestBase(), ArtifactIntegrationTest
     }
 
     @Test
+    fun `Builds apk on-the-fly under configuration cache`() {
+        // #1187: Cannot query the value of this provider because it has no value available.
+        val result = execute("", "publishReleaseApk", extraArgs = listOf("--configuration-cache"))
+
+        result.requireTask(":packageRelease", outcome = SUCCESS)
+        assertThat(result.output).contains("uploadApk(")
+        assertThat(result.output).contains(".apk")
+        assertThat(result.output).contains("publishArtifacts(")
+    }
+
+    @Test
     fun `Using custom artifact with single default mapping file uploads it`() {
         // language=gradle
         val config = """

